### PR TITLE
Feature/wav2vec2 updates

### PR DESCRIPTION
**Pull request by HalestormAI**

### CODE REVIEW

File: optimum/graphcore/models/wav2vec2/modeling_wav2vec2.py [hunk 1]
```diff
@@ -281,28 +308,189 @@ def forward(
...
         """
         Compute logits for contrastive loss based using cosine similarity as the distance measure between
         `[positive_feature, negative_features]` and `[predicted_features]`. Additionally, temperature can be applied.
         """
         target_features = torch.cat([target_features, negative_features], dim=0)
 
-        logits = torch.cosine_similarity(
-            predicted_features.float(), target_features.float(), dim=-1, eps=1e-4
-        ).type_as(target_features)
+        logits = torch.cosine_similarity(predicted_features, target_features, dim=-1, eps=1e-4)
```
Comment (HalestormAI) [inline optimum/graphcore/models/wav2vec2/modeling_wav2vec2.py:322]: Do we definitely not need the casts here?

File: examples/speech-pretraining/run_pretraining.py [hunk 2]
```diff
@@ -43,13 +44,12 @@
 from transformers.utils import check_min_version
 from transformers.utils.versions import require_version
 
-
 # Will error if the minimal version of Transformers is not installed. Remove at your own risks.
 check_min_version("4.15.0.dev0")
 
+# require_version("transformers==4.18.0", "To fix: pip install -r examples/speech-pretraining/requirements.txt")
```
[the remark this answers is not in the document]
Comment (thorinf) [reply examples/speech-pretraining/run_pretraining.py:50]: Maybe we can keep it in. I was looking for newer versions to try but they are unreleased.

File: examples/speech-pretraining/run_pretraining.py [hunk 3]
```diff
@@ -275,17 +275,26 @@ def __call__(self, features: List[Dict[str, Union[List[int], torch.Tensor]]]) ->
             min_masks=1,
         )
 
+        cropped_length = int(mask_indices_seq_length * self.model.config.mask_time_prob) + 1
+        # position true masked indexes first and crop, we mask the false indexes later
```
[the remark this answers is not in the document]
Comment (thorinf) [reply examples/speech-pretraining/run_pretraining.py:279]: # move true masked indexes first and crop, later used to gather a reduced size tensor.

I think this is a little better. Hard to word it nicely because it contains a small amount of the false positions too, but the true is first

File: examples/speech-recognition/README.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
+<!---
```
[the remark this answers is not in the document]
Comment (thorinf) [reply examples/speech-recognition/README.md:1]: # Copyright (c) 2020 Graphcore Ltd. All rights reserved.
 ?

File: examples/speech-recognition/run_inference_ctc.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+# take argmax and decode
+output = inference_model(**batch)
+logits = output[0]
+predicted_ids = torch.argmax(logits, dim=-1)
+transcription = processor.batch_decode(predicted_ids)
+print(transcription)
+
+# benchmark
+for i in tqdm(range(10000)):
```
[the remark this answers is not in the document]
Comment (thorinf) [reply examples/speech-recognition/run_inference_ctc.py:59]: I could add this. I was aiming to match the style of https://huggingface.co/facebook/wav2vec2-large-960h and the benchmark running was just stuck on afterwards.

File: optimum/graphcore/modeling_utils.py [hunk 6]
```diff
@@ -196,7 +195,7 @@ def recomputation_checkpoint(module: nn.Module) -> torch.utils.hooks.RemovableHa
     def recompute_outputs(module, inputs, outputs):
         if isinstance(outputs, torch.Tensor):
             return poptorch.recomputationCheckpoint(outputs)
-        elif isinstance(outputs, tuple):
+        else:
             return tuple(poptorch.recomputationCheckpoint(y) for y in outputs)
```
[the remark this answers is not in the document]
Comment (thorinf) [reply optimum/graphcore/modeling_utils.py:199]: Let me try without and i'll revert. I remember there was a bug with this originally which James was aware of. I've copied the code from my repo, but i'll double check

File: optimum/graphcore/models/wav2vec2/modeling_wav2vec2.py (hunk 1 above)
Comment (thorinf) [reply optimum/graphcore/models/wav2vec2/modeling_wav2vec2.py:322]: If its BASE we can put them back in. For LARGE there's no way it will compile without half-precision or a serious change in the loss forward-backward

File: examples/speech-recognition/base_4-ipu_config.json [hunk 7]
```diff
@@ -0,0 +1,27 @@
+{
+  "embedding_serialization_factor": 0,
+  "recompute_checkpoint_every_layer": false,
+  "optimizer_state_offchip": false,
+  "replicated_tensor_sharding": true,
+  "enable_half_first_order_momentum": true,
+  "enable_half_partials": true,
+  "device_iterations": 1,
+  "inference_device_iterations": 4,
```
[the remark this answers is not in the document]
Comment (thorinf) [reply examples/speech-recognition/base_4-ipu_config.json:9]: Yes. We do need to drop the last batch if its incomplete though, so if iters is higher for eval stage then its likely to miss more from the set - not a big problem really.

File: examples/speech-pretraining/large_8-ipu_config.json [hunk 8]
```diff
@@ -0,0 +1,35 @@
+{
+  "embedding_serialization_factor": 0,
+  "recompute_checkpoint_every_layer": false,
+  "optimizer_state_offchip": true,
+  "replicated_tensor_sharding": true,
+  "enable_half_first_order_momentum": true,
+  "enable_half_partials": true,
+  "device_iterations": 1,
+  "inference_device_iterations": 8,
```
[the remark this answers is not in the document]
Comment (thorinf) [reply examples/speech-pretraining/large_8-ipu_config.json:9]: Same as above

File: examples/speech-recognition/run_inference_ctc.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+model = AutoModelForCTC.from_pretrained("facebook/wav2vec2-base-960h")
+num_device_iterations = 10
+ipu_config = IPUConfig(inference_device_iterations=num_device_iterations)
+
+# processor = Wav2Vec2Processor.from_pretrained("facebook/wav2vec2-large-960h")
+# model = AutoModelForCTC.from_pretrained("facebook/wav2vec2-large-960h")
+# num_device_iterations = 2 # must be multiple of pipeline stages (2)
+# ipu_config = IPUConfig(matmul_proportion=0.1, inference_device_iterations=num_device_iterations,
+#                        layers_per_ipu=[17, 16],
+#                        ipus_per_replica=2)
```
[the remark this answers is not in the document]
Comment (thorinf) [reply examples/speech-recognition/run_inference_ctc.py:25]: I was aiming to match the style of https://huggingface.co/facebook/wav2vec2-large-960h for VoiceLab
Could be modified no problem

File: examples/speech-recognition/run_inference_ctc.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+sample_batch = {'input_values': torch.zeros([num_device_iterations, 320000])}
+
+inference_model.compile(**sample_batch)
+
+# load dummy dataset and read soundfiles
+ds = load_dataset("patrickvonplaten/librispeech_asr_dummy", "clean", split="validation")
+
+# get batch
+x = torch.zeros([num_device_iterations, 320000])
```
[the remark this answers is not in the document]
Comment (thorinf) [reply examples/speech-recognition/run_inference_ctc.py:42]: Just needs to be greater than LibriSpeechs longest, 20s adds a little extra

File: examples/speech-pretraining/README.md [hunk 11]
```diff
@@ -58,7 +58,7 @@ python run_pretraining.py \
 	--dataset_name "librispeech_asr" \
 	--dataset_config_name "clean" \
 	--train_split_name "validation" \
-	--ipu_config_name Graphcore/wav2vec2-base-ipu \
+	--ipu_config_name "./base_4-ipu_config.json" \
```
[the remark this answers is not in the document]
Comment (HalestormAI) [reply examples/speech-pretraining/README.md:61]: I think we should aim to do that, not urgently, but before we PR back to the main HF repo. Could you ask PSE how they did that?

File: examples/speech-pretraining/run_pretraining.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (HalestormAI) [reply examples/speech-pretraining/run_pretraining.py:279]: Yeah that's clearer.

File: examples/speech-recognition/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (HalestormAI) [reply examples/speech-recognition/README.md:1]: That's the one (but 2022).

File: optimum/graphcore/models/wav2vec2/modeling_wav2vec2.py (hunk 1 above)
Comment (HalestormAI) [reply optimum/graphcore/models/wav2vec2/modeling_wav2vec2.py:322]: In which case; I think this should be controlled by an argument so it can be switched in the config files.

File: examples/speech-recognition/run_inference_ctc.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (HalestormAI) [reply examples/speech-recognition/run_inference_ctc.py:59]: I see, in which case, perhaps remove the benchmark bit to match the example there.

**Review by HalestormAI — COMMENTED** [review]
LGTM - I'll close the PR. Thanks @thorinf
